### PR TITLE
Generalize PIFO Trees to PIEO Trees

### PR DESCRIPTION
**Pull request by polybeandip**
This PR makes the following changes
- switches from PIFOs to PIEOs
- alters our work-conserving policies to use immediately ready packets, thereby producing the same results on PIEO trees
- implements a few NWC policies via PIEO tree:
    + `FCFS(A,B)` but every packet is ready 5 seconds after it's pushed
    + `WFQ(A, B, C, D)` but flow `C` is rate limited to 78 bytes/s

### CODE REVIEW

**Comment by polybeandip** [comment]
**UPDATE**: I've noticed the graphs for `WFQ(A. B. C. D)`, with `C` rate limited, are not the same before and after compilation :(

For this reason, I'm turning this back into a draft. Apologies for any unnecessary pings

**Comment by polybeandip** [comment]
**UPDATE**: the graphs match now :)

File: lib/alg.ml [hunk 1]
```diff
@@ -520,17 +509,97 @@ module Alg2T (Alg : Alg_t) : Alg_t = struct
...
 
   let z' s pkt =
-    let pt, s' = Alg.control.z s pkt in
-    (f_tilde Alg.topology pt, s')
+    let pt, s', ts = Alg.control.z s pkt in
+    (f_tilde pt, s', ts)
 
   let control : Control.t =
-    { s = Alg.control.s; q = Pifotree.create topology; z = z' }
+    { s = Alg.control.s; q = Pieotree.create topology; z = z' }
```
Comment (anshumanmohan) [inline lib/alg.ml:519]: Nice, I'm thrilled that the changes in this file up to now were relatively straightforward. Thanks for tidying up my silly string-based matches into a little type of its own ("A" -> A), that's much nicer

File: lib/alg.ml [hunk 2]
```diff
@@ -520,17 +509,97 @@ module Alg2T (Alg : Alg_t) : Alg_t = struct
...
+      wfq_helper s weight var_last_finish (Packet.len pkt) time
+    in
+    let ts, s'' = 
+      if flow <> C then 
+        Time.epoch, s'
+      else
+        let c_ts = State.lookup "C_next_ts" s' in
+        let throttle = State.lookup "C_throttle" s' in
+        Time.of_float c_ts, 
+        State.rebind "C_next_ts" (Float.ceil (Packet.len pkt /. throttle) +. c_ts) s'
```
Comment (anshumanmohan) [inline lib/alg.ml:573]: Woohoo, terrific!!

File: lib/pieo.ml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    match peek t with
+    | None -> List.rev acc
+    | Some x ->
+        ignore (pop_exn t);
+        helper (x :: acc)
+  in
+  helper []
+
+let pop_pred t f = 
+  (* Pop the least element in the PIFO that satisfies f. *)
```
Comment (anshumanmohan) [inline lib/pieo.ml]: A few instances in this file where you say PIFO but mean PIEO

File: lib/pieotree.ml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  match (t, path) with
+  | Leaf p, [ (_, r) ] -> Leaf (Pieo.push p (pkt, r, ts))
+  | Internal (qs, p), (i, r) :: pt ->
+      let p' = Pieo.push p (i, r, ts) in
+      let q' = push (List.nth qs i) ts pkt pt in
+      Internal (Util.replace_nth qs i q', p')
+  | _ -> failwith "Push: invalid path"
+
+let rec size t now =
+  (* The size of a PIFO tree is the number of ready packets in its leaves.
```
Comment (anshumanmohan) [inline lib/pieotree.ml]:
```suggestion
  (* The size of a PIEO tree is the number of ready packets in its leaves. 
```

File: lib/pieotree.ml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  (* A leaf is well-formed.
+      An internal node is well-formed if:
+       - each of its child trees is well-formed
+       - the number of ready packets in the ith child-tree is equal to the number
+         of ready i's in _its own_ PIEO. 
+  *)
+  let pieo_count_occ p ele = 
+    Pieo.count p (fun (v, _, ts) -> v = ele && ts <= now) 
+  in
+  (* Counts how many times ele occurs as a value in PIFO p. *)
```
Comment (anshumanmohan) [inline lib/pieotree.ml]:
```suggestion
  (* Counts how many times ele occurs as a value in PIEO p. *)
```

**Review by anshumanmohan — APPROVED** [review]
I'm super duper pleased with this, well done!! Just a few little comments for you but you should feel free to merge after those are done. And, correct me if I'm wrong, but things seem to have basically proceeded as smoothly as hoped in https://github.com/cucapra/packet-scheduling/discussions/16?

If you don't mind, please take 15 mins at tomorrow's meeting to give us a demo of this work!

**Comment by polybeandip** [comment]
> And, correct me if I'm wrong, but things seem to have basically proceeded as smoothly as hoped in [cucapra/packet-scheduling#16](https://github.com/cucapra/packet-scheduling/discussions/16)?

Totally, everything went exactly as your discussion post predicted!

> If you don't mind, please take 15 mins at tomorrow's meeting to give us a demo of this work!

Sure!

**Pull request merged**